### ISSUE TITLE
java.lang.ClassCastException: android.os.BinderProxy cannot be cast to com.mapbox.android.telemetry.TelemetryService$TelemetryBinder

### ISSUE THREAD

**Issue by sheng930920** [mapbox/mapbox-gl-native#12460]
map sdk version: 6.2.1

implementation('com.mapbox.mapboxsdk:mapbox-android-sdk:6.2.1@aar') {
        transitive = true
    }

Manufacturer: Xiaomi
Model: Redmi Note 5A
CrashReporter Key: 9DF46B5E-2592-E00A-25CE-760DBEB66DB13DB66B66
Date: Mon Jul 23 18:58:25 GMT+08:00 2018

java.lang.ClassCastException: android.os.BinderProxy cannot be cast to com.mapbox.android.telemetry.TelemetryService$TelemetryBinder
at com.mapbox.android.telemetry.MapboxTelemetry$3.onServiceConnected(MapboxTelemetry.java:380)
at android.app.LoadedApk$ServiceDispatcher.doConnected(LoadedApk.java:1461)
at android.app.LoadedApk$ServiceDispatcher$RunConnection.run(LoadedApk.java:1489)
at android.os.Handler.handleCallback(Handler.java:754)
at android.os.Handler.dispatchMessage(Handler.java:95)
at android.os.Looper.loop(Looper.java:163)
at android.app.ActivityThread.main(ActivityThread.java:6377)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:904)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:794)

**Comment by LukasPaczos** [comment]
Thanks for reaching out and reporting the crash @sheng930920.

Should this one be covered by https://github.com/mapbox/mapbox-events-android/pull/180, @electrostat?

**Comment by electrostat** [comment]
That's correct. This is available in the latest Telemetry Snapshot and will be worked into an upcoming maps release. Let me know if you still experience this problem when testing the snapshot.

**Comment by sheng930920** [comment]
I am using the latest mapbox sdk version: 6.3.0, But the problem still exists. How should this problem be solved?

implementation 'com.mapbox.mapboxsdk:mapbox-android-sdk:6.3.0'


Android: 7.1.2
Manufacturer: Xiaomi
Model: Redmi Note 5A
CrashReporter Key: 9DF46B5E-2592-E00A-25CE-760DBEB66DB13DB66B66
Date: Tue Jul 24 10:56:34 GMT+08:00 2018

java.lang.ClassCastException: android.os.BinderProxy cannot be cast to com.mapbox.android.telemetry.TelemetryService$TelemetryBinder
at com.mapbox.android.telemetry.MapboxTelemetry$3.onServiceConnected(MapboxTelemetry.java:380)
at android.app.LoadedApk$ServiceDispatcher.doConnected(LoadedApk.java:1461)
at android.app.LoadedApk$ServiceDispatcher$RunConnection.run(LoadedApk.java:1489)
at android.os.Handler.handleCallback(Handler.java:754)
at android.os.Handler.dispatchMessage(Handler.java:95)
at android.os.Looper.loop(Looper.java:163)
at android.app.ActivityThread.main(ActivityThread.java:6377)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:904)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:794)

**Comment by sheng930920** [comment]
@electrostat Is there any temporary solution to this problem?

**Comment by sheng930920** [comment]
![image](https://user-images.githubusercontent.com/13199067/43115461-8dac9a3a-8f36-11e8-8c7a-4a0ba4f6cb52.png)

**Comment by sheng930920** [comment]
Only when the Android application starts multi-process, and the Android version is above 6.0, the problem will occur, and when the multi-process is removed, it will return to normal....

Recurrence is as follows:
(1) When the multi-process is started in the service of the Manifest.xml file, running the app crashes.
![image](https://user-images.githubusercontent.com/13199067/43116429-4cbe0e92-8f3a-11e8-9387-e1582854acb3.png)
![image](https://user-images.githubusercontent.com/13199067/43116436-547888e2-8f3a-11e8-824f-1c655da2a9c8.png)

**Comment by electrostat** [comment]
@sheng930920 it is not included in the Maps SDK 6.3.0. It'll be in an upcoming release. You need to exclude telemetry from the maps sdk dependency and include the telemetry snapshot. I'll post some instructions below.

//remove telemetry and core from the maps dependency, within your app's build.gradle
api(dep.mapboxMap) {
        transitive = true
        exclude group: 'com.mapbox.mapboxsdk', module: 'mapbox-android-telemetry'
        exclude group: 'com.mapbox.mapboxsdk', module: 'mapbox-android-core'
    }

//add the telemetry SNAPSHOT dependency to your build.gradle
api ''com.mapbox.mapboxsdk:mapbox-android-telemetry:3.2.0-SNAPSHOT'

Note: make sure "maven { url "http://oss.sonatype.org/content/repositories/snapshots/" }" is one of your repositories within `allprojects` of your projects build.gradle . and you see the snapshot in your list of external libraries. 

Test it again and we will know if this fix will work for you. Hope that helps. I'm not sure on the maps release timeline. Any note on the next release @LukasPaczos ?

**Comment by LukasPaczos** [comment]
The stable `6.4.0` is targeting mid-August. Before that, you will probably be able to test the fix with `6.4.0-alpha.2` that's going to be available next week, or the snapshot which @electrostat mentioned.

**Comment by Sarahhuang** [comment]
My application also has this problem after upgrading the version of maps SDK to 6.3.0.  I tried the above method to replace the telemetry version to 3.2.0-SNAPSHOT or replace the map SDK version to 6.4.0-alpha.1,  but it didn't work.  And I found that this problem only occurs on some models such as Samsung G9500, Samsung G930F, Nexus 5x  and Xiaomi MI 4LTE, but some do not appear such as Xiaomi MI 6, HUAWEI Mate 9. Later I found that this problem does not occur when I remove the JPush init method from my BaseApplication where I add the mapbox init method. It seems that mapbox's init method has some conflict with Jpush's init method. But both of them work well when I use the maps SDK version 5.3.0.

**Comment by Sarahhuang** [comment]
My application also has this problem after upgrading the version of maps SDK to 6.3.0.

I have tried the above methods to replace the telemetry version to **3.2.0-SNAPSHOT** or replace the map SDK version to **6.4.0-alpha.1**, but they didn't work. And I found that this problem only occurs on some models such as Samsung G9500, Samsung G930F, Nexus 5x and Xiaomi MI 4LTE, but some do not appear such as Xiaomi MI 6, HUAWEI Mate 9. 

Later I found that this problem does not occur when I remove the JPush init method from my BasicApplication where I add the mapbox init method. 

It seems that mapbox's init method has some conflict with Jpush's init method. But both of them work well when I use the maps SDK 5.3.1.

```java
public class BasicApplication extends MultiDexApplication {

    @Override
    public void onCreate() {
        super.onCreate();
        //JPushInterface.init(this);  //remove this init method
        Mapbox.getInstance(this, "xxxxxxxx");
    }
 }
```

**Comment by Sarahhuang** [comment]
JPush's PushService  defaults in ":pushcore" process in its library manifest  and TelemetryService is in default process, both of them bind a service in their init method.   When binding across process boundaries, an instance of BinderProxy is used instead of the actual instance that causes the crash.  Finally, I solved this problem by replacing the process of  PushService‘s process in my main manifest.

```xml
<service android:name="cn.jpush.android.service.PushService"
            android:enabled="true"
            android:exported="false"
            tools:node="replace"> 
           <!--Add this line to replace the library manifest with this node.-->
            <intent-filter>
                <action android:name="cn.jpush.android.intent.REGISTER" />
                <action android:name="cn.jpush.android.intent.REPORT" />
                <action android:name="cn.jpush.android.intent.PushService" />
                <action android:name="cn.jpush.android.intent.PUSH_TIME" />
            </intent-filter>
        </service>
```

**Comment by 3mph4515** [comment]
Also faced with this annoying problem. Excluding telemetry, as @electrostat mentioned doesn't help.

**Comment by ben-j69** [comment]
I also still have this issue in 6.40...

Only 5.5.3 is ok

**Comment by ben-j69** [comment]
I even tried the 6.5.0-alpha.1

`08-28 11:52:56.345 14600-14600/com.wayzup.wayzupapp.debug:location_service E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.wayzup.wayzupapp.debug:location_service, PID: 14600
    java.lang.ClassCastException: android.os.BinderProxy cannot be cast to com.mapbox.android.telemetry.TelemetryService$TelemetryBinder
        at com.mapbox.android.telemetry.MapboxTelemetry$3.onServiceConnected(MapboxTelemetry.java:380)
        at android.app.LoadedApk$ServiceDispatcher.doConnected(LoadedApk.java:1335)
        at android.app.LoadedApk$ServiceDispatcher$RunConnection.run(LoadedApk.java:1352)
        at android.os.Handler.handleCallback(Handler.java:739)
        at android.os.Handler.dispatchMessage(Handler.java:95)
        at android.os.Looper.loop(Looper.java:158)
        at android.app.ActivityThread.main(ActivityThread.java:7224)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1230)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1120)`

**Comment by CoderDaMing** [comment]
I also still have this issue in com.mapbox.mapboxsdk:mapbox-android-sdk:6.4.0

**Comment by electrostat** [comment]
I have finally been able to replicate this bug. It's an issue when multiple services are attempting to bind to the application context. Likely caused by a developer using their own service or an additional sdk they are using. They are running on different processes and the second one (ours in this case) is sending back a BinderProxy, which we can not use. I will investigate further and come up with a solution that can address this.

**Comment by pavelpoley** [comment]
I am facing this issue on 6.5.0, started crashing when i added Yandex metrica (from manifest i see they run service on another process) , also with multidex enabled

**Comment by LukasPaczos** [comment]
@pavelpoley this issue has been fixed in the `v6.6.0`. Let us know if you're still running into it after upgrading.
